### PR TITLE
ARM application gateway with Autorest implemented

### PR DESCRIPTION
**Pull request by MikhailTryakhov**
ARM application gateways commands implemented on swagger

### CODE REVIEW

File: lib/commands/arm/network/appGateway._js [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+      appGateway.sku.name = options.skuName;
+    }
+    if (options.skuTier) {
+      utils.verifyParamExistsInCollection(constants.appGateway.sku.tier, options.skuTier, '--sku-tier');
+      appGateway.sku.tier = options.skuTier;
+    }
+    if (options.capacity && capacity < constants.sku.capacity[0] && capacity > constants.sku.capacity[1]) {
+      appGateway.sku.capacity = options.capacity;
+    } else {
+      //TODO: add another case functionality
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: Please remove or implement

File: lib/commands/arm/network/appGateway._js [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+
+    self.interaction.formatOutput(appGateways, function (data) {
+      if (data.length === 0) {
+        self.output.warn(util.format($('No application gateways found in resource group "%s"'), resourceGroup));
+      } else {
+        self.output.table(data, function (row, gateway) {
+          row.cell($('Name'), gateway.name);
+          row.cell($('Provisioning state'), gateway.provisioningState);
+          row.cell($('Location'), gateway.location);
+          row.cell($('Resource group'), self._getResourceGroupFromId(gateway));
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: please use `resourceUtils.js` instead

File: lib/commands/arm/network/appGateway._js [hunk 3]
```diff
@@ -0,0 +1,923 @@
...
+      return;
+    }
+
+    var progress = self.interaction.progress(util.format($('Deleting an application gateway "%s"'), appGatewayName));
+    try {
+      self.networkManagementClient.applicationGateways.deleteMethod(resourceGroup, appGatewayName, _);
+    } finally {
+      progress.end();
+    }
+    self.list(resourceGroup, options, _);
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: No need to list AG after delete, please remove

File: lib/commands/arm/network/appGateway._js [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+      throw new Error(util.format($('Application gateway "%s" not found in resource group "%s"'), appGatewayName, resourceGroup));
+    }
+
+    var frontendPort = utils.findFirstCaseIgnore(appGateway.frontendPorts, {name: frontendPortName});
+    if (frontendPort) {
+      throw new Error(util.format($('A frontend port with name "%s" already exists for an application gateway "%s"'), frontendPortName, appGatewayName));
+    }
+    frontendPort = {
+      name: frontendPortName,
+      port: parseInt(options.port)
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: Can we add port validation here?

File: lib/commands/arm/network/appGateway._js [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+    var rule = {
+      name: ruleName,
+      type: constants.appGateway.routingRule.type[0],
+      backendHttpSettings: backendHttpSettings,
+      httpListener: httpListener,
+      backendAddressPool: backendAddressPool
+    };
+
+    if (options.type) {
+      rule.type = options.type;
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: Can we add type validation here?

File: lib/commands/arm/network/appGateway._js [hunk 6]
```diff
@@ -0,0 +1,923 @@
...
+        namesString += ', ';
+      }
+      namesString += item.name;
+      counter++;
+    });
+    namesString += ']';
+    return namesString;
+  },
+
+  _getResourceGroupFromId: function (appGateway) {
```
Comment (romangromov) [inline lib/commands/arm/network/appGateway._js]: Can we use ` resourceUtils.getResourceInformation` instead?

File: lib/commands/arm/network/constants._js [hunk 7]
```diff
@@ -92,6 +92,26 @@ module.exports = {
...
+    },
+    settings: {
+      affinity: ['Disabled', 'Enabled'],
+      protocol: ['Http'],
+      port: [0, 65535],
+      defPort: 80
+    },
+    sku: {
+      capacity: [2, 10],
+      name: ['Standard_Small', 'Standard_Medium', 'Standard_Large'],
```
Comment (romangromov) [inline lib/commands/arm/network/constants._js]: This should be generated dynamically

File: lib/commands/arm/network/network._js [hunk 8]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .description($('Commands to manage application gateways'));
+
+  appGateway.command('create [resource-group] [name] [location]')
+    .description($('Create an application gateway'))
+    .usage('[options] <resource-group> <name> <location>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-n, --name <name>', $('the name of the application gateway'))
+    .option('-l, --location <location>', $('the location'))
+    .option('-e, --vnet-name <vnet-name>', $('the name of the virtual network application gateway should be deployed in'))
+    .option('-m, --subnet-name <subnet-name>', $('the name of subnet in the virtual network identified by --vnet-name'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: There should be `--subnet-id` parameter to specify subnet id directly

File: lib/commands/arm/network/network._js [hunk 9]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+      constants.appGateway.settings.protocol))
+    .option('-o, --http-settings-port <http-settings-port>', util.format($('the HTTP settings port, valid range is'),
+      utils.toRange(constants.appGateway.settings.port)))
+    .option('-f, --http-settings-cookie-based-affinity <http-settings-cookie-based-affinity>',
+      util.format($('Enable or disable HTTP settings cookie based affinity, valid values are' +
+      '\n     [%s],' +
+      '\n     default value is "%s"'), constants.appGateway.settings.affinity, constants.appGateway.settings.affinity[0]))
+    .option('-j, --frontend-port <frontend-port>', util.format($('the frontend port value, valid range is'),
+      utils.toRange(constants.appGateway.settings.port)))
+    .option('-k, --public-ip-name <public-ip-name>', $('the name of the public ip'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: There should be `--publicip-id` parameter to specify public ip id directly

File: lib/commands/arm/network/network._js [hunk 10]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-w, --routing-rule-type <routing-rule-type>', util.format($('the request routing rule type, default is "%s"'), constants.appGateway.routingRule.type[0]))
+    .option('-a, --sku-name <sku-name>', util.format($('the name of the sku, valid values are \[%s\]. Default values is "%s"'),
+      constants.appGateway.sku.name, constants.appGateway.sku.name))
+    .option('-t, --sku-tier <sku-tier>', util.format($('the sku tier, valid values are \[%s\]. Default values is "%s"'),
+      constants.appGateway.sku.name.tier, constants.appGateway.sku.name.tier))
+    .option('-z, --capacity <capacity>', util.format($('application gateway instance count in range \[%s\]. Default value is %s.'),
+      constants.appGateway.sku.capacity, constants.appGateway.sku.capacity[0]))
+    .option('-d, --description <description>', $('the description for the application gateway'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: Is there any restriction on description length?

File: lib/commands/arm/network/network._js [hunk 11]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-s, --subscription <subscription>', $('the subscription identifier'))
+    .option('-t, --tags <tags>', $(constants.help.tags.create))
+    .execute(function (resourceGroup, name, location, options, _) {
+      resourceGroup = cli.interaction.promptIfNotGiven($('Resource group name: '), resourceGroup, _);
+      name = cli.interaction.promptIfNotGiven($('Application gateway name: '), name, _);
+      location = cli.interaction.promptIfNotGiven($('Location: '), location, _);
+      options.vnetName = cli.interaction.promptIfNotGiven($('Virtual network name: '), options.vnetName, _);
+      options.subnetName = cli.interaction.promptIfNotGiven($('Virtual network subnet name: '), options.subnetName, _);
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: we should be able to provide subnet id directly

File: lib/commands/arm/network/network._js [hunk 12]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .description($('Set an application gateway'))
+    .usage('[options] <resource-group> <name>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-n, --name <name>', $('the name of the application gateway'))
+    .option('-l, --sku-name <sku-name>', $('the name of the sku'))
+    .option('-t, --sku-tier <sku-tier>', $('the sku tier'))
+    .option('-z, --capacity <capacity>', util.format($('application gateway instance count in range \[%s\]. Default value is %s.'),
+      constants.appGateway.sku.capacity, constants.appGateway.sku.capacity[0]))
+    .option('-d, --description <description>', $('the description for the application gateway'))
+    .option('-s, --subscription <subscription>', $('the subscription identifier'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: we should be able to unset existing tags, like we do for any kind of resources

File: lib/commands/arm/network/network._js [hunk 13]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+      appGateway.create(resourceGroup, name, location, options, _);
+    });
+
+  appGateway.command('set [resource-group] [name]')
+    .description($('Set an application gateway'))
+    .usage('[options] <resource-group> <name>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-n, --name <name>', $('the name of the application gateway'))
+    .option('-l, --sku-name <sku-name>', $('the name of the sku'))
+    .option('-t, --sku-tier <sku-tier>', $('the sku tier'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: Can we add valid range for sku-tier in description?

File: lib/commands/arm/network/network._js [hunk 14]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-z, --capacity <capacity>', util.format($('application gateway instance count in range \[%s\]. Default value is %s.'),
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: There should be no default value for `set` command, right?

File: lib/commands/arm/network/network._js [hunk 15]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+      name = cli.interaction.promptIfNotGiven($('Application gateway name: '), name, _);
+
+      var networkManagementClient = getNetworkManagementClient(options);
+      var appGateway = new AppGateway(cli, networkManagementClient);
+      appGateway.set(resourceGroup, name, options, _);
+    });
+
+  appGateway.command('list [resource-group]')
+    .description($('List resource group application gateways'))
+    .usage('[options] <resource-group>')
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: This wrong: `resource-group` is now optional, not required

File: lib/commands/arm/network/network._js [hunk 16]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-s, --subscription <subscription>', $('the subscription identifier'))
+    .execute(function (resourceGroup, name, options, _) {
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: if not provided, then `resourceGroup` becomes `undefined` here, possible bug

File: lib/commands/arm/network/network._js [hunk 17]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+      var networkManagementClient = getNetworkManagementClient(options);
+      var appGateway = new AppGateway(cli, networkManagementClient);
+      appGateway.list(resourceGroup, options, _);
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: the same as above: `resource-group` is now optional

File: lib/commands/arm/network/network._js [hunk 18]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+  var appGatewaySslCert = appGateway.category('ssl-cert')
+    .description($('Commands to manage application gateway SSL certificates'));
+
+  appGatewaySslCert.command('add [resource-group] [gateway-name] [name]')
+    .description($('Add application gateway SSL certificate'))
+    .usage('[options] <resource-group> <gateway-name> <name>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-n, --gateway-name <gateway-name>', $('the name of the application gateway'))
+    .option('-c, --name <name>', $('the name of the certificate'))
+    .option('-f, --cert-file <cert-file>', $('the path to the certificate'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: which types of certificates allowed here?

File: lib/commands/arm/network/network._js [hunk 19]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-p, --password <password>', $('the certificate password'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: Is there any restrictions on password?

File: lib/commands/arm/network/network._js [hunk 20]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-s, --subscription <subscription>', $('the subscription identifier'))
+    .execute(function (resourceGroup, gatewayName, name, options, _) {
+      resourceGroup = cli.interaction.promptIfNotGiven($('Resource group name: '), resourceGroup, _);
+      gatewayName = cli.interaction.promptIfNotGiven($('Application gateway name: '), gatewayName, _);
+      name = cli.interaction.promptIfNotGiven($('Certificate name: '), name, _);
+      options.certFile = cli.interaction.promptIfNotGiven($('Certificate file path: '), options.certFile, _);
+      options.password = cli.interaction.promptIfNotGiven($('Certificate password: '), options.password, _);
+
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: certFile and password  are required parameters, can we declare this in command description?

File: lib/commands/arm/network/network._js [hunk 21]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .description('Commands to manage application gateway ip config');
+
+  appGatewayIpConfig.command('set [resource-group] [gateway-name] [name]')
+    .description($('Update an ip config to an application gateway'))
+    .usage('[options] <resource-group> <gateway-name> <name>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-w, --gateway-name <gateway-name>', $('the name of the application gateway'))
+    .option('-n, --name <name>', $('the name of the application gateway ip config'))
+    .option('-e, --vnet-name <vnet-name>', $('the name of the virtual network'))
+    .option('-u, --subnet-name <subnet-name>', $('the name of the subnet'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: we should be able to provide subnet id directly

File: lib/commands/arm/network/network._js [hunk 22]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+
+  appGatewayFrontendIp.command('add [resource-group] [gateway-name] [name]')
+    .description($('Add a frontend ip configuration to an application gateway'))
+    .usage('[options] <resource-group> <gateway-name> <name>')
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-w, --gateway-name <gateway-name>', $('the name of the application gateway'))
+    .option('-n, --name <name>', $('the name of the frontend IP address name'))
+    .option('-e, --vnet-name <vnet-name>', $('the name of the virtual network'))
+    .option('-u, --subnet-name <subnet-name>', $('the name of the subnet'))
+    .option('-i, --subnet-id <subnet-id>', $('the id of the subnet'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: Yes, that is correct. Can we add `<...>-id` parameters everywhere?

File: lib/commands/arm/network/network._js [hunk 23]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-a, --static-ip-address <static-ip-address>', $('the static IP address name'))
+    .option('-p, --public-ip-name <public-ip-name>', $('the name of the public ip name'))
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: we should be able to provide publicip id directly

File: lib/commands/arm/network/network._js [hunk 24]
```diff
@@ -2265,6 +2266,479 @@ exports.init = function (cli) {
...
+    .option('-s, --subscription <subscription>', $('the subscription identifier'))
+    .execute(function (resourceGroup, gatewayName, name, options, _) {
+      resourceGroup = cli.interaction.promptIfNotGiven($('Resource group name: '), resourceGroup, _);
+      gatewayName = cli.interaction.promptIfNotGiven($('Application gateway name: '), gatewayName, _);
+      name = cli.interaction.promptIfNotGiven($('Frontend IP name: '), name, _);
+
+      if (options.vnetName || options.subnetName) {
```
Comment (romangromov) [inline lib/commands/arm/network/network._js]: can we place this logic inside `appGateway.addFrontendIp`

**Comment by romangromov** [comment]
@MikhailTryakhov please see general comments from my end:

* Please add `<...>-id` parameters (subnet-id, publicip-id, etc.) for all of dependencies, like we do everywhere in Xplat-CLI - there should be ability to specify resource `ID` directly;
* Please add **tests** for all of new commands implemented as well as **nocks** recordings;
* Please **sync** your branch with upstream and resolve conflicts.

Q: do we have any kind of **quality tests**  for this new set of commands? If no, can we share it with AG team to get feedback at first before pushing to production code? Thanks

File: lib/commands/arm/network/constants._js (hunk 7 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/constants._js]: Could you please add a bit more details here - do we have examples of getting available ranges dinamically?

Comment (romangromov) [inline lib/commands/arm/network/constants._js]: @MikhailTryakhov the SKUs `name` field is just a concatenation of its `tier` and `size`. Instead, we can specify SKUs `tier` and `size` to generate `name` programatically, like we do in `ExpressRoute Circuits`

Comment (MikhailTryakhov) [inline lib/commands/arm/network/constants._js]: cool, but guys from AG team asked to do it in the same style as Powhershell commands - we have no size or family parameters there, just use constant values Standard_Small | Standard_Medium | Standard_Large
Please review it [here](https://msdn.microsoft.com/en-us/library/mt619435.aspx)

Comment (romangromov) [inline lib/commands/arm/network/constants._js]: @MikhailTryakhov ok, no action required if confirmed from AG

File: lib/commands/arm/network/network._js (hunk 19 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Not found

File: lib/commands/arm/network/network._js (hunk 24 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: hm... do you really think that we should move prompt logic to local functions? I assume that would be useful to input all parameters in network files.
the other way - exclude prompts and throw error like 'you should enter subnet or public ip to create frontend ip'

Comment (romangromov) [inline lib/commands/arm/network/network._js]: let us prompting instead of throwing error. This is just recommendation, like we do in another commands

**Comment by bingosummer** [comment]
It seems that arm app gw in Azure CLI does not support [Custom health probe](https://azure.microsoft.com/en-us/documentation/articles/application-gateway-probe-overview/).

**Comment by MikhailTryakhov** [comment]
Hi @bingosummer 
sorry for delay. Currently we've migrated app gw code on swagger without probes. But was not pubished yet.
I assume we'll publish it on 0.9.15 with probes - they're in progress

**Comment by romangromov** [comment]
@MikhailTryakhov 
There is support for `AG probes` inside autorest SDK:
```
 * @param {array} [parameters.probes] Gets or sets probes of application
 * gateway resource
```
Can we include `probes` in this PR as well?

**Comment by MikhailTryakhov** [comment]
@romangromov correct. As I told above, they're in progress and would be included

File: lib/commands/arm/network/network._js (hunk 8 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Resolved

File: lib/commands/arm/network/network._js (hunk 9 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Resolved

File: lib/commands/arm/network/network._js (hunk 21 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Resolved

File: lib/commands/arm/network/network._js (hunk 22 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Resolved

File: lib/commands/arm/network/network._js (hunk 23 above)
Comment (MikhailTryakhov) [inline lib/commands/arm/network/network._js]: Resolved

**Comment by MikhailTryakhov** [comment]
Newest version would be pushed later

**Comment by bingosummer** [comment]
@MikhailTryakhov 
Could you please update the current status of the latest PR? Is is ready?

**Comment by romangromov** [comment]
@bingosummer let's start new issue at this repo ("ARM: App Gateways commands not available"), add reference to this PR and track progress on this

**Comment by MikhailTryakhov** [comment]
Hi @bingosummer current progress is in another branch - this one is outdated

**Pull request closed without merging**